helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -1636,4 +1648,54 @@ synchronized void closeRebalancer() {
       }
     }
   }
+
+  Timer _participantDeregistrationTimer = null;
+  AtomicReference<ParticipantDeregistrationTask> _nextParticipantDeregistrationTask = new AtomicReference<>();
+  class ParticipantDeregistrationTask extends TimerTask {
```

Comment:
We can still leverage the original task with a specified time. It does not have to be compute logic in the task. It can be part of the pipeline stage. Also, you can use the an util "Schedule Pipeline" something. It is in the code base somewhere.

Reply:
Left a question in your above comment. Agree that we do not need a new task + timer if we move participant deregistration to be part of the rebalancer. 

My view is that increasing the frequency of onDemand rebalances will add more work to the cluster rather than having a separate pipeline for participant deregistration. There may also be a lot of on demand rebalances that are called just for deregistering a participant, which will then trigger its own rebalance. 

Are the main benefits of moving deregistration stage to the rebalance pipeline reducing code and keeping ZK writes grouped to end of rebalancer? Thank you for the advice

Reply:
Unify the zk write stage and minimize the write is the major consideration. 

But the more important is to reduce the heaviness of the code. For this simple pipeline trigger and we create a "Task" class is too much.